Fix mosdepth coverage in general stats if key is missing

jchorl opened the pull request:
Fixes: https://github.com/MultiQC/MultiQC/issues/3431

If e.g. 80% of reads have 50x coverage, and we want to show 40x coverage, we know at least 80% of reads have 40x coverage. This is particularly important in general stats where there is no "connecting line" chart. So if we don't get the exact right coverage increment that is stored, we end up with an empty box in general stats.

Tested locally:

Before:

<img width="1240" height="155" alt="2025-12-10_16-23" src="https://github.com/user-attachments/assets/63404f70-5d5c-4dd9-b54e-3280dd2a34cc" />


After:

<img width="1194" height="170" alt="2025-12-10_16-22" src="https://github.com/user-attachments/assets/6708b8ff-7f21-40b7-8cef-08a63d5f3fe2" />

ewels commented:
This is definitely better than showing 0% (which is wrong) but it's still showing incorrect data, which could be a little misleading.

Any ideas how we could flag when this is happening? At least I think we should add to the module docstring so that the behaviour ends up in the docs.

jchorl commented:
You know, we don't even use mosdepth for other reasons. We just output coverage in a format compatible with mosdepth formatting. I don't really understand why this optimization is the default in mosdepth - it's complex, to save at most a few mb (I think - unless other folks use mosdepth in other ways).

I just updated our code to not apply this optimization. Now we just output coverage at every threshold and avoid this bug altogether. I'm going to close this. I don't necessarily think third-party apps should need to work around this behaviour in mosdepth, it should probably be defaulted off upstream.

ewels commented:
I'm happy for this PR to continue, I think it's generally useful for folks. If it does get fixed upstream then the code in this PR won't apply, so it won't do any harm.

I'd just like a note in the docs about it basically, then I think it's good to merge.

jchorl commented:
Added a note to the docs! Of note, I did test this and added tests. However we don't run mosdepth internally and the mosdepth logic is quite complex, so there is some risk I got something wrong. I think values should only be missing at >300x coverage so impact is probably small.

ewels reviewed (approved):
Thanks for this. Happy Christmas!
